Set VAULT_ADDR in process env when using token helper

ianferguson opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-vault/blob/master/.github/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
### Description

Closes #526 

I'm opening this as a draft pull request because I'd like to get the maintainers thoughts on the approach being taken.

I've been using a token helper that stores tokens based on the configured value of VAULT_ADDR in the environment while using terraform across many Vault clusters, and having support for this would make it much easier to run changes en masse and avoid needing to manage setting and unsetting the VAULT_ADDR for different applies based on what was configured in the Vault terraform provider.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-vault/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
VAULT_ADDR is always set to the provider's address attribute when a Vault token helper is used to resolve a Vault token.
```

Output from acceptance testing:

- None at this time -- if the general approach is ok with the maintainers, I can write up some specific tests for this.

tyrannosaurus-becks commented:
Thank you for your PR. I have thoughts and questions on it and to avoid splitting the conversation, I'm going to comment on the linked issue further.

ianferguson commented:
@tyrannosaurus-becks sounds good! I was offline all last week, so will check in on the discussion on the original issue, thanks for following up on this

ianferguson commented on vault/provider_test.go:
```diff
@@ -116,8 +116,7 @@ func getTestRMQCreds(t *testing.T) (string, string, string) {
 // A basic token helper script.
-const tokenHelperScript = `
-#!/usr/bin/env bash
+const tokenHelperScript = `#!/usr/bin/env bash
```
The extra newline added after the backtick caused the script to come out as 

```

#!/usr/bin/env bash
echo "helper-token"
```

With the shebang on the 2nd, rather than 1st, line. When executed as an exec by the cmd package for a user using fish shell the follow error silently happened:

```
> /usr/local/bin/fish -c $tmpScriptPath
Failed to execute process '$tmpScriptPath'. Reason:
exec: Invalid argument
The file '$tmpScriptPath' is marked as an executable but could not be run by the operating system.
```

Removing the line break here fixes that otherwise perplexing test failure that non-bash users may see

ianferguson commented on vault/provider_test.go:
```diff
@@ -387,17 +386,8 @@ func TestAccProviderToken(t *testing.T) {
-		}()
-	}
-	err = os.Unsetenv(config.ConfigPathEnv)
+	reset, err := tempUnsetenv(config.ConfigPathEnv)
```
I ended up using TestAccProviderToken as a template for writing TestAccProviderVaultAddEnv, and in the process of reading them and writing another similar test consolidated some of the setup/tear down refactoring -- if this is not ok I can unwind the changes here

ianferguson commented:
@tyrannosaurus-becks I've updated the PR with a flag (`add_address_to_env`) for turning the new functionality on and off (and it defaults to the current non-adding behavior), added tests and updated the documentation.

The tests pass for me against a local `vault server -dev`, but fail on circleci with:
```
--- FAIL: TestAccProviderVaultAddrEnv (0.02s)
    --- FAIL: TestAccProviderVaultAddrEnv/AddAddressToEnvNotConfigured (0.01s)
        provider_test.go:683: bad token value: want "https://pretest-env-var.example.com", got "http://127.0.0.1:8200"
    --- FAIL: TestAccProviderVaultAddrEnv/AddAddressToEnvIsFalse (0.01s)
        provider_test.go:683: bad token value: want "https://pretest-env-var.example.com", got "http://127.0.0.1:8200"
    --- FAIL: TestAccProviderVaultAddrEnv/AddAddressToEnvIsTrue (0.00s)
        provider_test.go:683: bad token value: want "https://provider.example.com", got "http://127.0.0.1:8200"
```

Which I'll take a look at on Monday or Tuesday, but otherwise I think this should be about ready for review based on what we discussed on #526

ianferguson commented on vault/provider_test.go:
```diff
@@ -640,3 +600,206 @@ func testTokenName_check(expectedTokenName string) resource.TestCheckFunc {
+	// when the ExternalTokenHelper's command is formatted into a shell invocation to exec here:
+	// https://github.com/hashicorp/vault/blob/master/command/token/helper_external.go#L117-L132
+	//
+	resetBashEnvEnv, err := tempUnsetenv("BASH_ENV")
```
I think its possible that a better fix here might be to move all of the vars set here: https://github.com/terraform-providers/terraform-provider-vault/blob/f42716aae3aebc8daf9702dfa20ce3f8d09d9f4d/.circleci/config.yml#L26-L32 to an `Environment` object in the circleci yaml configuration for the container these tests are run in, as documented here: https://circleci.com/docs/2.0/env-vars/#setting-an-environment-variable-in-a-container

That being said, I don't know the finer details of these tests or how circle manages those environment variables/containers across multiple steps well enough to be confident that that wouldn't have any unintended side effects, so I took a more surgical route to fixing these tests that rely on testing the VAULT_ADDR change when a token helper is called.
